File: algorithm/Leetcode/app/src/main/java/com/cxmax/third/arrays/BinarySearch.java
package com.cxmax.third.arrays;

/**
 * 三刷
 * <p>
 * https://leetcode.cn/problems/binary-search/
 * <p>
 * Created by caixi on 2022/7/19.
 */
public class BinarySearch {

    /**
     * 左闭右闭
     * 1.确定左右，位数
     * 2.遍历
     * 3.写条件找middle
     *
     * @param nums
     * @param target
     * @return
     */
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while(left <= right) {
            int mid = left + ((right - left) >> 1);
            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                return mid;
            }
        }
        return -1;
    }

}

File: algorithm/Leetcode/app/src/main/java/com/cxmax/selftest/arrays/BinarySearch.kt
package com.cxmax.selftest.arrays

/**
 *
 *  https://leetcode-cn.com/problems/binary-search/
 *
 *  704.二分查找
 *
 * Created by caixi on 2021/12/27.
 */
object BinarySearch {

    fun test() {
        val array = arrayOf(0, 1, 2, 3, 4)
        binarySearch(array, 3)
    }

    /**
     * 采用左闭右闭的写法
     * @param array Array<Int>
     * @return Int
     */
    fun binarySearch(array: Array<Int>?, target: Int): Int {
        if (array == null || array.size == 0) {
            return -1
        }
        var left = 0
        var right = array.size - 1
        while (left <= right) {
            val middle = left + ((right - left) / 2)
            if (array[middle] < target) {
                left = middle + 1
            } else if (array[middle] > target) {
                right = middle - 1
            } else {
                return middle
            }
        }
        return -1
    }
}